src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs

```diff
@@ -0,0 +1,194 @@
+
+    public ShardedStudioCollectionsHandlerProcessorForPreviewCollection(ShardedRequestHandler requestHandler)
+        : base(requestHandler)
+    {
+    }
+
+    protected override void Initialize()
+    {
+        base.Initialize();
+
+        _releaseContext = RequestHandler.ContextPool.AllocateOperationContext(out _context);
+
+        _continuationToken = RequestHandler.GetOrCreateContinuationToken(_context);
+    }
+
+    protected override JsonOperationContext GetContext()
+    {
+        return _context;
+    }
+
+    protected override long GetTotalResults()
+    {
+        return _result.TotalResults;
+    }
+
+    protected override bool NotModified(out string etag)
+    {
+        DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Karmel, DevelopmentHelper.Severity.Normal, "Need to figure out the best way to combine ETags and send not modified");
+
+        etag = null;
+        return false;
+    }
+
+    protected override async ValueTask<List<Document>> GetDocumentsAsync()
+    {
+        var op = new ShardedCollectionPreviewOperation(RequestHandler, _continuationToken, _context);
+        _result = await RequestHandler.ShardExecutor.ExecuteParallelForAllAsync(op);
+
+        return _result.Results;
+    }
+
+    protected override List<string> GetAvailableColumns(List<Document> documents)
+    {
+        return _result.AvailableColumns;
+    }
+
+    public override void Dispose()
+    {
+        base.Dispose();
+
+        _releaseContext?.Dispose();
+        _releaseContext = null;
+    }
+
+    private class PreviewCollectionResult
+    {
+        public List<Document> Results;
+        public long TotalResults;
+        public List<string> AvailableColumns;
+    }
+
+    private readonly struct ShardedCollectionPreviewOperation : IShardedOperation<PreviewCollectionResult>
+    {
+        private readonly ShardedRequestHandler _handler;
+        private readonly ShardedPagingContinuation _token;
+        private readonly JsonOperationContext _context;
+
+        public ShardedCollectionPreviewOperation(ShardedRequestHandler handler, ShardedPagingContinuation token, JsonOperationContext context)
+        {
+            _handler = handler;
+            _token = token;
+            _context = context;
+        }
+
+        public PreviewCollectionResult Combine(Memory<PreviewCollectionResult> results)
+        {
+            var availableColumns = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
+            var total = new PreviewCollectionResult
+            {
+                Results = new List<Document>()
+            };
+
+            var span = results.Span;
+            var totalDocuments = 0L;
+
+            for (int i = 0; i < span.Length; i++)
+            {
+                availableColumns.AddRange(span[i].AvailableColumns);
+                total.TotalResults += span[i].TotalResults;
+            }
+
+            for (int i = 0; i < span.Length * _token.PageSize; i++)
```

Comment:
Why do you multiple by the page size here?

Reply:
Wouldn't it be better to simply have a `while` loop that takes from each of those in turn?

Reply:
This all is a quick & dirty solution.

It is a common pattern which probably will be use in multiple places. 
So would like to rework it to reuse `MergedEnumerator` and that we will stream the results from the shards rather than just materialize them all.